### PR TITLE
Distributed demo tutorial

### PR DESCRIPTION
**Pull request by chester-leung**

### CODE REVIEW

File: demo/python/distributed/distr-training.py [hunk 1]
```diff
@@ -7,11 +7,6 @@
 
 enclave = xgb.Enclave(HOME_DIR + "build/enclave/xgboost_enclave.signed")
 
-# Remote Attestation
```
Comment (podcastinator) [inline demo/python/distributed/distr-training.py:10]: Should we actually include attestation in this demo, instead of removing it?

File: doc/tutorials/distributed.rst [hunk 2]
```diff
@@ -4,6 +4,105 @@ Distributed Secure XGBoost
...
+* `Distributed Training`_
+
+* `Tutorial`_
+
+* `Troubleshooting`_
+
+************
+Architecture
+************
+Distributed Secure XGBoost leverages a tracker, or an orchestrator, to configure the cluster. This involves setting up the cluster topology, i.e. the communication patterns. Depending on the data size, the tracker may either set up the nodes in a tree pattern or a ring pattern.
```
Comment (podcastinator) [inline doc/tutorials/distributed.rst]: 1. Tracker vs. orchestrator can be confusing. I think we should use a single term and stick with it.
2. "i.e. the communication patterns. Depending on the data size..." --> "i.e., the communication patterns: depending on the data size..."

File: doc/tutorials/distributed.rst [hunk 3]
```diff
@@ -4,6 +4,105 @@ Distributed Secure XGBoost
...
+.. figure:: images/ring.png
+   :width: 100% 
+   :alt: Ring Topology 
+   :align: center
+
+Once the topology has been set up, each node in the cluster is assigned a rank. The rank is essentially the node identifier in the cluster. The node assigned rank 0 is considered the "master" node, and all other nodes are considered worker nodes.
+
+The tracker is logically distinct from all nodes in the cluster, and does not actually perform any computation. Once it has finished setting up the topology, the tracker has finished its job. However, a user may choose to colocate the tracker with a master/worker node for simplicity. 
+
+Before training begins, Distributed Secure XGBoost leverages *inter-enclave attestation* to authenticate all enclaves in the cluster -- each enclave attests all neighboring enclaves to verify the enclave's identity and to verify that the enclave will be running the proper code.
```
Comment (podcastinator) [inline doc/tutorials/distributed.rst]: Should we add something here about what the client attests?

File: doc/tutorials/distributed.rst [hunk 4]
```diff
@@ -4,6 +4,105 @@ Distributed Secure XGBoost
...
+      13.95.157.223:22
+      40.68.135.193:22
+
+2. Ensure that ``distr-training.py`` is identical and at the same path on every machine in the cluster. 
+
+3. Start computation. This will start the tracker and subsequently the job.
+
+   .. code-block:: bash
+
+      mc2-xgboost/host/dmlc-core/tracker/dmlc-submit --cluster ssh --host-file hosts.config --num-workers <num_nodes_in_cluster> --worker-memory 4g python3 distr-training.py
```
Comment (podcastinator) [inline doc/tutorials/distributed.rst:90]: Should we say something here about how to check the list of arguments / options?

File: doc/tutorials/distributed.rst [hunk 5]
```diff
@@ -4,6 +4,105 @@ Distributed Secure XGBoost
...
+
+
+   Ensure that each node in the cluster has authorized the SSH public key of the machine running the tracker, as the tracker in Distributed Secure XGBoost leverages SSH public keys to set up the topology.
```
Comment (podcastinator) [inline doc/tutorials/distributed.rst:104]: I think we should introduce the SSH-based connections earlier, perhaps in the architecture...

**Review by podcastinator — COMMENTED** [review]
This looks great, thanks! Left some comments.

File: demo/python/distributed/distr-training.py (hunk 1 above)
Comment (chester-leung) [reply demo/python/distributed/distr-training.py:10]: `xgb.rabit.init()` does inter-enclave attestation right? Deleting this will make it such that there's no explicit attestation, but there's still attestation happening in the background.

File: doc/tutorials/distributed.rst (hunk 2 above)
Comment (chester-leung) [reply doc/tutorials/distributed.rst]: Fixed

File: doc/tutorials/distributed.rst (hunk 3 above)
Comment (chester-leung) [reply doc/tutorials/distributed.rst]: There's no client in this demo. It's just a demo of distributed training -- there's no concept of client/server as we're not doing outsourced computation.

File: doc/tutorials/distributed.rst (hunk 4 above)
Comment (chester-leung) [reply doc/tutorials/distributed.rst:90]: I added something detailing each argument.

File: doc/tutorials/distributed.rst (hunk 5 above)
Comment (chester-leung) [reply doc/tutorials/distributed.rst:104]: Fixed

File: demo/python/distributed/distr-training.py (hunk 1 above)
Comment (podcastinator) [reply demo/python/distributed/distr-training.py:10]: Ah but the client still needs to attest the cluster, or at least the "master" enclave...

Comment (chester-leung) [reply demo/python/distributed/distr-training.py:10]: In this demo there's no concept of a client. This is similar to the secure-xgboost-demo, in that there's no concept of client/server. Everything is happening on the server, or in this case, a cluster.

Comment (podcastinator) [reply demo/python/distributed/distr-training.py:10]: Yes but even in the secure-xgboost-demo, we still invoke the remote attestation APIs to show what they look like (also in the e2e-demo). I'm imagining something like that for this demo as well.

File: doc/tutorials/distributed.rst (hunk 3 above)
Comment (podcastinator) [reply doc/tutorials/distributed.rst]: See comment above.

File: doc/tutorials/distributed.rst [hunk 6]
```diff
@@ -38,6 +38,8 @@ Once the topology has been set up, each node in the cluster is assigned a rank.
 
 The tracker is logically distinct from all nodes in the cluster, and does not actually perform any computation. Once it has finished setting up the topology, the tracker has finished its job. However, a user may choose to colocate the tracker with a master/worker node for simplicity. 
 
+The tracker initially communicates with all nodes in the cluster over SSH. This means that the tracker's SSH public key must be authorized by every node in the cluster. Once this initial communication has finished, all other inter-node communication happens over TLS.
```
Comment (podcastinator) [inline doc/tutorials/distributed.rst]: Instead of saying the tracker communicates with the other nodes over SSH, I think it might be clearer to rephrase this to something like: the tracker first logs into each node in the cluster via SSH to spin up the enclaves.
Then, merge this para with the next and say that the enclaves attest each other, and while doing so establish TLS connections with each other.

File: doc/tutorials/distributed.rst [hunk 7]
```diff
@@ -91,6 +93,17 @@ If you are using Azure Confidential Computing, all your VMs must be on the same
       mc2-xgboost/host/dmlc-core/tracker/dmlc-submit --cluster ssh --host-file hosts.config --num-workers <num_nodes_in_cluster> --worker-memory 4g python3 distr-training.py
 
 
+   The command takes in the following arguments:
+
+         * ``--cluster`` : how the cluster is set up. In Secure XGBoost we leverage SSH for intra-cluster communication.
```
Comment (podcastinator) [inline doc/tutorials/distributed.rst:95]: Do we wanna list the possible values for this argument? e.g. `ssh` vs `local`. 
Or better still, is there a help command that dumps all this info?

Comment (chester-leung) [reply doc/tutorials/distributed.rst:95]: We could list it. There should be a help command, but all the possible arguments are also listed here: https://github.com/dmlc/dmlc-core/blob/master/tracker/dmlc_tracker/opts.py. The 4 I listed are really all we need though, so I'm not sure if it'd be helpful to list all possible arguments and their possible values.

I'll add a note about a local cluster though.

File: doc/tutorials/distributed.rst (hunk 6 above)
Comment (chester-leung) [reply doc/tutorials/distributed.rst]: Fixed

File: demo/python/distributed/distr-training.py (hunk 1 above)
Comment (chester-leung) [reply demo/python/distributed/distr-training.py:10]: Added this back in

File: doc/tutorials/distributed.rst [hunk 8]
```diff
@@ -38,11 +38,8 @@ Once the topology has been set up, each node in the cluster is assigned a rank.
...
 
-The tracker initially communicates with all nodes in the cluster over SSH. This means that the tracker's SSH public key must be authorized by every node in the cluster. Once this initial communication has finished, all other inter-node communication happens over TLS.
-
-Before training begins, Distributed Secure XGBoost leverages *inter-enclave attestation* to authenticate all enclaves in the cluster -- each enclave attests all neighboring enclaves to verify the enclave's identity and to verify that the enclave will be running the proper code.
+Before training begins, the tracker first logs into each node in the cluster via SSH and configures it for distributed training. Distributed Secure XGBoost then leverages *inter-enclave attestation* to authenticate all enclaves in the cluster -- each enclave attests all neighboring enclaves to verify the enclave's identity and to verify that the enclave will be running the proper code. During attestation, enclaves establish TLS connections with one another, enabling a secure channel to communicate on whie training.
```
Comment (podcastinator) [inline doc/tutorials/distributed.rst]: Typo: whie --> while

Comment (chester-leung) [reply doc/tutorials/distributed.rst]: Argh. Fixed!

**Review by podcastinator — APPROVED** [review]
